core/src/main/java/kafka/server/share/SharePartitionManager.java

```diff
@@ -606,22 +641,51 @@ private void maybeCompleteInitializationWithException(
             return;
         }
 
+        // Remove the partition from the cache as it's failed to initialize.
+        partitionCacheMap.remove(sharePartitionKey);
+        // The partition initialization failed, so complete the request with the exception.
+        // The server should not be in this state, so log the error on broker and surface the same
+        // to the client. The broker should not be in this state, investigate the root cause of the error.
+        log.error("Error initializing share partition with key {}", sharePartitionKey, throwable);
+        maybeCompleteShareFetchExceptionally(future, Collections.singletonList(topicIdPartition), throwable);
+    }
+
+    private void handleSharePartitionException(
+        SharePartitionKey sharePartitionKey,
+        Throwable throwable
+    ) {
         if (throwable instanceof NotLeaderOrFollowerException || throwable instanceof FencedStateEpochException) {
             log.info("The share partition with key {} is fenced: {}", sharePartitionKey, throwable.getMessage());
             // The share partition is fenced hence remove the partition from map and let the client retry.
             // But surface the error to the client so client might take some action i.e. re-fetch
             // the metadata and retry the fetch on new leader.
             partitionCacheMap.remove(sharePartitionKey);
```

Comment:
Does the `SharePartition` need to be put into a fenced state? Removing it from the cache is good for future requests, but is that enough for the object which is already in existence?

Reply:
So we remove share partition from cache at 2 places. 1) When initialization failed 2) When fenced error occurs.
For 1, it's safe as its still in initilization state.

For 2, I was in mixed opinion. As all interactions with share partition happens currently while fetching instance from cache hence once removed or re-initialized the new state should appear. But if old share partition instance is already held by some other thread then `acknowledge` will anyways fail but `fetch` can succeed. Do you think it would be sensible to have another state in `SharePartition` as `Fenced`, which once set then `fetch lock` on that share partition cannot be attained. Do you think we should have an `active` status check on all Share Partition APIs as well?
cc: @adixitconfluent

Reply:
I think its better to have a new state. Should we have a state as `TERMINATED`/`FENCED` because a share partition should also be removed from SPM in case the topic partition is deleted or becomes a follower. JIRA - https://issues.apache.org/jira/browse/KAFKA-17783. Wdyt @apoorvmittal10 @AndrewJSchofield

Reply:
Yes, I think these states sound sensible.

Reply:
> Do you think we should have an active status check on all Share Partition

Yes, I agree on this as well

Reply:
Do we have a state machine written down somewhere for SharePartition?

Reply:
Hmmm, not yet. Let me create the state transitions better in a single place where we can see the transitions.